cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+   Free Software Foundation; version 3.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <package_module.h>
+#include <pipes.h>
+#include <signals.h>
+#include <buffer.h>
+#include <string_lib.h>
+#include <actuator.h>
+#include <file_lib.h>
+#include <known_dirs.h>
+#include <locks.h>
+#include <rlist.h>
+#include <policy.h>
+#include <eval_context.h>
+
+#define INVENTORY_LIST_BUFFER_SIZE 100 * 80 /* 100 entries with 80 characters 
+                                             * per line */
+
+static bool UpdateSinglePackageModuleCache(EvalContext *ctx,
+                                    const PackageModuleWrapper *module_wrapper,
+                                    UpdateType type, bool force_update);
+static char *GetPackageModuleRealPath(const char *package_manager_name);
+static int NegotiateSupportedAPIVersion(PackageModuleWrapper *wrapper);
+
+
+void DeletePackageModuleWrapper(PackageModuleWrapper *wrapper)
+{
+    free(wrapper->path);
+    free(wrapper->name);
+    free(wrapper);
+}
+
+PackageModuleWrapper *NewPackageModuleWrapper(PackageModuleBody *package_module)
+{
+    assert(package_module && package_module->name);
+
+    PackageModuleWrapper *wrapper = xmalloc(sizeof(PackageModuleWrapper));
+    
+    wrapper->path = GetPackageModuleRealPath(package_module->name);
+    wrapper->name = SafeStringDuplicate(package_module->name);
+    wrapper->package_module = package_module;
+    
+    /* Check if file exists */
+    struct stat sb;
+    if (!wrapper->path || (stat(wrapper->path, &sb) != 0))
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "can not find package wrapper in provided location: %s",
+            wrapper->path);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    
+    /* Negotiate API version */
+    wrapper->supported_api_version = NegotiateSupportedAPIVersion(wrapper);
+    if (wrapper->supported_api_version != 1)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "unsupported package module wrapper API version: %d",
+            wrapper->supported_api_version);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    return wrapper;
+}
+
+
+void UpdatePackagesCache(EvalContext *ctx, bool force_update)
+{
+    Log(LOG_LEVEL_DEBUG, "Updating package cache.");
```

Comment:
cf-agent/cf-agent.c:255 describes this as a package **promise** cache - which is it ?